engine/graphics/src/opengl/graphics_opengl.cpp

```diff
@@ -1751,54 +1815,153 @@ static uintptr_t GetExtProcAddress(const char* name, const char* extension_name,
         {
 #ifndef NDEBUG
             GLint logLength;
-            glGetShaderiv(s, GL_INFO_LOG_LENGTH, &logLength);
+            glGetShaderiv(shader_id, GL_INFO_LOG_LENGTH, &logLength);
             if (logLength > 0)
             {
                 GLchar *log = (GLchar *)malloc(logLength);
-                glGetShaderInfoLog(s, logLength, &logLength, log);
+                glGetShaderInfoLog(shader_id, logLength, &logLength, log);
                 dmLogError("%s\n", log);
                 free(log);
             }
 #endif
-            glDeleteShader(s);
+            glDeleteShader(shader_id);
             return 0;
         }
 
-        return s;
+        return shader_id;
+    }
+
+    static Shader* CreateShader(GLenum type, ShaderDesc::Shader* ddf)
+    {
+        // Note: The DDF will be removed after the shader has been created
+        //       so we need to store a copy of the texture array source
+        //       since the programs that use this variant will patch in the
+        //       maximum number of pages supported before creating the 
+        //       shader program.
+        if (ddf->m_VariantTextureArray)
+        {
+            Shader* shader = new Shader();
+            shader->m_VariantTextureArrayData     = (char*) malloc(ddf->m_Source.m_Count);
+            shader->m_VariantTextureArrayDataSize = ddf->m_Source.m_Count;
+            memcpy(shader->m_VariantTextureArrayData, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+
+            for (int i = 0; i < ddf->m_Uniforms.m_Count; ++i)
+            {
+                if (ddf->m_Uniforms[i].m_Type == ShaderDesc::SHADER_TYPE_SAMPLER2D_ARRAY)
+                {
+                    shader->m_VariantTextureArrayUniforms.OffsetCapacity(1);
+                    shader->m_VariantTextureArrayUniforms.Push(strdup(ddf->m_Uniforms[i].m_Name));
+                }
+            }
+
+            return shader;
+        }
+        else
+        {
+            GLuint shader_id = DoCreateShader(type, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+            if (!shader_id)
+            {
+                return 0;
+            }
+            Shader* shader = new Shader();
+            shader->m_Id = shader_id;
+            return shader;
+        }
     }
 
     static HVertexProgram OpenGLNewVertexProgram(HContext context, ShaderDesc::Shader* ddf)
     {
-        assert(ddf);
-        return CreateShader(GL_VERTEX_SHADER, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+        return (HVertexProgram) CreateShader(GL_VERTEX_SHADER, ddf);
     }
 
     static HFragmentProgram OpenGLNewFragmentProgram(HContext context, ShaderDesc::Shader* ddf)
     {
-        assert(ddf);
-        return CreateShader(GL_FRAGMENT_SHADER, ddf->m_Source.m_Data, ddf->m_Source.m_Count);
+        return (HFragmentProgram) CreateShader(GL_FRAGMENT_SHADER, ddf);
+    }
+
+    static GLuint CreateShaderVariantTextureArray(OpenGLProgram* program_ptr, Shader* shader, GLenum type, uint8_t max_page_count)
```

Comment:
This patches the generated shader source by replacing the DM_MAX_PAGE_COUNT macro with the actual max page count. It's not great that we have to do all this work, but as previously stated I don't know what the best option here is..